### DISCUSSION TITLE
How does KEDA handles downscaling?

### DISCUSSION THREAD

**Discussion by smintz** [apache/airflow#24067] [Q&A]
Hey There,
We want to start utilizing autoscaling with KEDA (we are about to install a fresh new cluster using the official helm chart).

I have concerns in regards to how KEDA handles downscaling. The scenario is simple: KEDA decides we have too many pods running, when set with `StatefulSet` or `Deployment`, Kubernetes will pick a pod to kill (based on metrics like utilization and stuff - but from our standpoint, this is as good as randomly select one). If this pod is not really idle, there might be a case were a job is running for 1h, at this point we will kill it and will have to restart it on another pod... not so good.

The other option is to use the KubernetesWorker which will spawn Kubernetes Jobs rather than Deployment/StatefulSet, which is imho the right thing to do. Problem is: spawning a job might take some time (sometime we will have to boot a new instance in order to get capacity).

I have two solutions in mind and would love to get feedback:
1. Use the KubernetesExecutor, but set some "placeholder" pods to allocate instances. The pods will use the same image as the jobs but will run `sleep infinity` and will have a lower priority, so an upcoming job will kill the placeholder and will be able to start faster.

2. Configure the CeleryWorker to exit after N tasks, and spawn it using KEDA, but with JobExecuter instead Deployment/StatefulSet. Alter the query to multiply by `1.1` (in order to have 10% spare capacity). This way, we will never kill a pod while its running a production task and will always have enough workers.

Would love to hear the feedback from the community.

**Comment by potiuk** [comment]
Yep. downscaling is an issue, and there are no good simple solutions yet. I believe @thesuperzapper worked on some solutions and there are some approaches that require modified versions of K8S to work.

**Reply by smintz** [reply]
Thanks for your answer @potiuk, is there a way to make a CeleryWorker exit if idle for X amount of time, or limit the number of jobs to accept, then exit?

**Reply by potiuk** [reply]
No idea.

**Comment by pingzh** [comment]
@smintz scaling down is tough. in our infra with celery executor, we only do weekly manual scaling down by stopping the celery worker and let the tasks run until finish (but with a max ttl). I am wondering what's your cluster scale?

**Comment by thesuperzapper** [comment]
@smintz I am working on a "task aware autoscaler" for celery workers that will hopefully be part of the next release of the [`User-Community Airflow Helm Chart`](https://github.com/airflow-helm/charts/tree/main/charts/airflow). It's coming along nicely but needs a bit more work before I share it publicly. 

(Follow this issue to see when it's done: https://github.com/airflow-helm/charts/issues/339)

__The main features will be:__

- Fully Standalone (no need for KEDA or anything except Kubernetes itself)
    - _(It runs as a Python script using SQLAlchemy, so actually could be contributed as a new "component" of airflow itself in the future)_
- Autoscaling of multiple different queues of celery workers
    - _(Allows things like having different Python dependencies between queues, or more/less resources, but not having workers from unused queues left running all day)_
- When downscaling, chooses the "best" workers to remove based on the configured allowance for disruption to tasks.
   - _(Uses the `pod-deletion-cost` annotation added in K8S 1.22 to influence which Pods are removed on downscale)_
- Some of the configs:
    - `target load factor`: the worker load-factor (percentage of total task slots used) to try and achieve, can be above 100% (for overprovisioning), or under 100% (to always have a task-slot buffer)
    - `allow removing active workers`: if active workers (with tasks running) are ever allowed to be downscaled, or if we must wait for one to be idle
    - `max wasted time`: the maximum amount of elapsed task-time that can be "wasted" by killing a worker
